### PR TITLE
[WIP] GPG Verification (#425)

### LINKED ISSUE #425: GPG signature verification
feature request/discussion:

Github just added GPG signature verification, this would be a cool feature in gogs as well.
https://github.com/blog/2144-gpg-signature-verification

ported from https://github.com/gogits/gogs/issues/2981

I think it's a realy nice to have feature

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/40238085-gpg-signature-verification?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bkcsoft**
Migrated from https://github.com/gogits/gogs/pull/3127

https://github.com/gogits/gogs/pull/3127#issuecomment-228626156
> I didn't tackle the part of the verification and I didn't have time to go further but i will re-put myself on it on the next days. Doing it in the front-end could be done easely. It just need the API to get the user public key from gogs like https://developer.github.com/v3/users/keys/. On the other hand, i don't thinks it would be too much work for the backen to do it. I have to think about it but at first at will implement the API because in both case they will be needed (or at least usefull)

Closes #425 

@sapk would you like to continue this by yourself? or mind filling me in on whats left to be done 🙂

### CODE REVIEW

**Comment by lunny** [comment]
Go ahead! I like this feature!

File: models/gpg_key.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/gogits/gogs/modules/log"
+)
+
+//TODO maybe refector with ssh-key ?
+//TODO database behind
+
+// PublicGPGKey represents a GPG key.
+type PublicGPGKey struct {
+	ID          int64  `xorm:"pk autoincr"`
+	OwnerID     int64  `xorm:"INDEX NOT NULL"`
```
Comment (lunny) [inline models/gpg_key.go:19]: If every user only has one GPG key?

Comment (bkcsoft) [reply models/gpg_key.go:19]: humm? AFAIK this allows for a one-to-many setup, so a single user can have many keys.

**Comment by ekozan** [comment]
you need to port this one too https://github.com/gogits/git-module/pull/13

**Comment by sapk** [comment]
I have done in git-module needed change to extract the sign information of the commit. In the gogs, I have done change to make a base to start with api route and db struct. From what I remember, I didn't make change to the ui yet and the validation of the sign by the key stored in gogs user of the commit.

Still to be done is : 
- Some missing api part (for example adding GPG key)
- UI to add GPG keys
- Validation of commit against users keys.

Most of what is missing should be allready commented as //TODO in code.

I will have time in three weeks to finish this but any help before is gladly accepted and i can assist to give information ^^

**Comment by lunny** [comment]
@spak would you like to continue this PR since you have some time and you are the original author? We can close this one if you will do it.

**Comment by sapk** [comment]
Gladly, in three week I will have time dedicated to contribute to a open-source project on my school time so I thinks this PR could be finish in that context.

**Comment by bkcsoft** [comment]
Closing in favour of #710

**Pull request closed without merging**